### ISSUE TITLE
Design Idea: the autarky score

### ISSUE THREAD

**Issue by Urwumpe** [mars-sim/mars-sim#132]
As another way to measure the progress in the simulation, I would suggest the autarky score. 

The simplest way to calculate it would be tracking how much of an item or amount resource is consumed and how much of the demand gets produced in a certain scope - overproduction would be capped to 100% autarky.

To measure this autarky properly, there would have to be a way to measure work done by settlers as well. 

(And maybe it would make sense to include the births on Mars as well in the calculation... but I believe it would be better suited for another statistic.)

There could be three good scopes to measure this autarky:

- Global/Mars
- Administrative region
- Settlement

Administrative region would be a new entity to keep track of new small outposts  and depots that can themselves never be autark but allow bigger settlements to achieve autarky easier. 

To avoid confusing jitter in the statistic, this value should be calculated for a long enough interval. I would suggest to present the data for the following intervals:

- Past orbit
- Current running orbit
- Any of the past 12 months. 

A more accurate calculation would keep track if a good is produced on Earth or on Mars or in a settlement, but that would require too big changes to the Inventory system to work. Also it isn't done that way on current Earth in macro-economics  as well, only inside big companies they keep track of every article by the ERP system. Sure it could be assumed that we have a standard ERP on mars as well, but it wouldn't be needed for the statistic. 

This autarky could also be used by the sponsor system as well for their possible decisions: A sponsor focussing on exploration would weight autarky less than one aiming at colonization.

**Comment by mokun** [comment]
@Urwumpe , 

Autark means autonomy, right ? That's very awesome you put forth the concept of independence or autonomy. 

> A more accurate calculation would keep track if a good is produced on Earth or on Mars or in a settlement, but that would require too big changes to the Inventory system to work. Also it isn't done that way on current Earth in macro-economics as well, 

If we are talking about economic autonomy, there is a rudimentary supply and demand system in mars-sim. It's based on what's called `Value Point` (VP). 

To know what VP is, let's also define what a `Good` is in mars-sim. 

It's an item that can be traded between settlements. A `Good` can be an `Amount Resource`, an `Item Resource`, `Equipment` or `Vehicle`. Each good has a VP that is changing over time.

> Note: as a matter of fact, each building Type also has a VP assigned to it, even though it is not a `Good` per se.

We have been using [GoodsManager](https://github.com/mars-sim/mars-sim/blob/master/mars-sim-core/src/main/java/org/mars_sim/msp/core/structure/goods/GoodsManager.java) class to update the VP of each Good a few times a day.

I'm thinking we could tweak `GoodsManager` to give out a composite score that measure the overall economic health of a settlement.

**Comment by mokun** [comment]
> Administrative region would be a new entity to keep track of new small outposts and depots that can themselves never be autark but allow bigger settlements to achieve autarky easier.

Yes we can make a `Mining Outpost` (or `Trading Outpost`) belong to a more full-blown settlement that help starting it. 

Currently, an outpost have its own VPs on the resources it carries. It does make more sense to integrate its supply and demand chain to that of a large settlement.

**Comment by mokun** [comment]
I just put together a new wiki called [Economics](https://github.com/mars-sim/mars-sim/wiki/Economics) as a reference. 

We can certainly improve and refine those underlining concepts created by previous developers since the current system has some limitations. 

For instance, it doesn't give the fair value of a resource.  The Value Points (VP) is just the demand and supply of a resource and is NOT the "true worth" of a resource.

e.g. each kg of O2 can worth 500,000 VP if its production is severely limited in a settlement. In comparison, a Cargo Rover may be only worth 500 VP.

**Comment by Urwumpe** [comment]
Yeah, if I understand it correctly, it could happen in manufacturing that combining parts worth 5000 VP could result in a good just 5 VP worth... of course, primitive capitalists would now cringe in pain, but Mars isn't an atomic market yet.

**Comment by Urwumpe** [comment]
BTW - what is the definition of an equipment in MSP?

**Comment by mokun** [comment]
>  manufacturing that combining parts worth 5000 VP could result in a good just 5 VP worth... of course, primitive capitalists would now cringe in pain, but Mars isn't an atomic market yet.

Luckily we don't add the VPs of all input resources of a manufacturing process to obtain the VP of the output resources 

The way we calculate the supply and demand is on per resource basis. 

> demand = # of times requested / kg

On the other hand, 

> supply = # of times produced / kg

**Comment by mokun** [comment]
> what is the definition of an equipment in MSP?

It's really whatever objects that extends the [Equipment](https://github.com/mars-sim/mars-sim/blob/master/mars-sim-core/src/main/java/org/mars_sim/msp/core/equipment/Equipment.java) class. 

So far, Bag, Barrel, EVASuit, GasCanister, LargeBag, SpecimenContainer, BuildingKit, and Robot are the ones.

BuildingKit is the latest addition. We haven't done anything with it yet.

**Comment by Urwumpe** [comment]
So, for example a Tablet/PDA like tool for a settler could also be a Equipment? 

I just wonder what the difference ist between Equipment and item resource is in the end, for example for inventory handling.

**Comment by mokun** [comment]
> what the difference ist between Equipment and item resource is in the end, for example for inventory handling.

It's up to us to define what a tablet should be. 

I would first define a tablet as a [Part](https://github.com/mars-sim/mars-sim/blob/master/mars-sim-core/src/main/resources/conf/parts.xml) and to see how well it works.

Currently, a `Part` belongs to [ItemResource](https://github.com/mars-sim/mars-sim/blob/master/mars-sim-core/src/main/java/org/mars_sim/msp/core/resource/ItemResource.java).

On the other hand, O2 is an [AmountResource](https://github.com/mars-sim/mars-sim/blob/master/mars-sim-core/src/main/java/org/mars_sim/msp/core/resource/AmountResource.java) to begin with and there's no question about it.

Initially, all these class (Bag, Barrel, EVASuit, GasCanister, LargeBag, SpecimenContainer) extend the [Container](https://github.com/mars-sim/mars-sim/blob/master/mars-sim-core/src/main/java/org/mars_sim/msp/core/equipment/Container.java) class, which in turns extend `Equipment`.

Only in the recent years, did we add Robot and BuildingKit to the Equipment class.

**Comment by Urwumpe** [comment]
So it is not like for example, an Equipment being a named unique entity in the simulation, while a Part or ItemResource is one identical clone out of many?

**Comment by mokun** [comment]
> Equipment being a named unique entity in the simulation, while a Part or ItemResource is one identical clone out of many?

If I examine what UNIQUENESS means. 

Say, there are two small hammers, one used by David and one used by Tracy. 

Both hammers would look the same and the Inventory class that store them won't be able to tell the differences between the two hammers.

So if an item is a Part or ItemResource, then it can NOT be UNIQUE. 

On the other hand, if an item is an equipment, then it can have its own name/designation.

e.g. EVA suit 001 is worn by David while EVA suit 002 is worn by Tracy. The two suits are two separate items.

An unique EVA suit object is created. 

Yes it's true that if we want a tablet to be defined as an equipment, then it will be UNIQUE.

**Comment by Urwumpe** [comment]
OK, now it starts to make more sense to me. Thank you. 👍

**Comment by mokun** [comment]
I was gonna say that using an unique object each bag, barrel, gas canister, etc. seems like a waste when there are hundreds of them.

So even though bag 001 and bag 002 may hold different resources inside, we should be able to come up with an alternative data structure such as a map to store resources, instead of having two instances of the Bag class.

Let's doing some refactoring !

**Comment by Urwumpe** [comment]
Maybe implement them close to a flyweight pattern - the bag equipment stores only the meta-data, but the actual inventory is handled by a special `MasterInventory` in `Simulation`. If you store water in the bag, the bag would refer to an amount of a single resource there. If you store a number of screws there, it would point to single type of item resources.

Maybe we can even get the refactoring then so far, that the pretty blob-like `Inventory` class right now gets turned into a plain facade to simplify working with different small, specialized and more efficient inventories, that maybe be created as needed for the content, instead of having one inventory that can store everything.

And maybe this could then get extended to the point, that loose heaps of objects are no longer common... everything is stored somewhere in something. Oxygen for the base in oxygen tanks, water in water tanks, EVA suits in lockers, empty bags in a box of bags of bags...

**Comment by mokun** [comment]
> A more accurate calculation would keep track if a good is produced on Earth or on Mars

Yes. We do need that as a way to gauge how well things are doing. 

For now, I'll add a simple HashMap data structure in Settlement class for tracking everything ever produced in ManufactureGood and ManufactureConstructionMaterials class for each sol.

**Comment by mokun** [comment]
I came across this particular definition of what a **Total Productivity** of a nation is.

![productivity](https://user-images.githubusercontent.com/1168584/53061101-9b1f9d80-3471-11e9-82fa-7dc34b526ad9.jpg)

Supposedly, the unit of all both unit and output is *Per Labor Hour*. So productivity will be essentially a ratio and is unit-less.

**Comment by Urwumpe** [comment]
Not automatically - the definition of productivity is just output per input. Daily productivity would be output per day / input per day, etc. 

And yes, its dimensionless - its a KPI for optimization. Just think of a settlement requiring to produce 8000 bricks and wants to choose the most productive process for that, then you could compare different processes by the productivity regardless of the time needed - a longer process could have less loss of input material and require less consumables along the process = more productive in that case.

**Comment by mokun** [comment]
> definition of productivity is just output per input. Daily productivity would be output per day / input per day, etc.

We have some of these stats or metrics computed in `Settlement` class prepared at the end of a sol. 

But that was when we were experimenting with the [MarqueeTicker](https://github.com/mars-sim/mars-sim/blob/master/mars-sim-ui/src/main/java/org/mars_sim/msp/ui/swing/tool/MarqueeTicker.java) class back then. 

In terms of UI design, it is a running bar (like a stock exchange news ticker) near the bottom of the *Settlement Map Tool* showing the major resource's usage over a period of a few sols.

However, I abandoned that effort after a while. It uses JIDE's StyledLabel and is quite limited in programming the news ticker. Also, it's in conflict of Weblaf library if I recall correctly. 

```
import com.jidesoft.swing.MarqueePane;
import com.jidesoft.swing.StyleRange;
import com.jidesoft.swing.StyledLabel;
```

Of course, your KPI is one step on top of that and yes we can implement to give players some ideas how well the settlement is running.

**Comment by mokun** [comment]
> We have some of these stats or metrics computed in `Settlement` class prepared at the end of a sol.
> 
> But that was when we were experimenting with the [MarqueeTicker](https://github.com/mars-sim/mars-sim/blob/master/mars-sim-ui/src/main/java/org/mars_sim/msp/ui/swing/tool/MarqueeTicker.java) class back then.

@bevans2000 @Urwumpe 
I was reminded that in every strategy game in that there's always a graph showing how much each resource would go up and down over time. 

We could do something similar if not the same.

Anyway I went ahead and revised the codes for generating some stats in this [commit](
https://github.com/mars-sim/mars-sim/commit/4d300aa956bcc0ae053dd69d814e2702bc05e5f8). 

For now, it outputs the average amount of 10 critical resources from yestersol in the settlement banner on the top part of the Settlement Map.
